### PR TITLE
Move NodeIPAMController to legacy-cloud-provider/gce

### PR DESCRIPTION
**Pull request by cici37**
**What type of PR is this?**
/kind cleanup

**What this PR does / why we need it**:
Google seems to be the only one who consume cloud cidr allocator hence it would be better to move to gcp specific place. The plan decided in cloud provider extraction meeting would be first to move to legacy-cloud-provider/gce. If there is no issue, it will be moved to cloud-provider-gcp repo eventually.

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

**Comment by cheftako** [comment]
/hold
Waiting to resolve https://github.com/kubernetes/kubernetes/issues/97835.

**Comment by cici37** [comment]
There is an recent issue raised about the possibility to expose the node CIDR allocator interface for out-of-tree cloud providers: https://github.com/kubernetes/kubernetes/issues/97835. Will hold until it is getter more clear

File: staging/src/k8s.io/legacy-cloud-providers/gce/nodeipam/config/v1alpha1/conversion.go [hunk 1]
```diff
@@ -19,7 +19,7 @@ package v1alpha1
 import (
 	"k8s.io/apimachinery/pkg/conversion"
 	"k8s.io/kube-controller-manager/config/v1alpha1"
```
Comment (jpbetz) [inline staging/src/k8s.io/legacy-cloud-providers/gce/nodeipam/config/v1alpha1/conversion.go:21]: Is `kube-controller-manager` still where `v1alpha1.NodeIPAMControllerConfiguration` should belong? `cloud-provider-gcp` depends on `kube-controller-manager` so this works out mechanically.

@cheftako are there any non-obvious constraints that we should be considering here? I checked the dependency tree and didn't find anything obviously problematic.

(same thing applies to defaults.go, which depends on the same config)

**Comment by aojea** [comment]
> Google seems to be the only one who consume cloud cidr allocator hence it would be better to move to gcp specific place.

The IPAM controller has two types of allocators

https://github.com/kubernetes/kubernetes/blob/5b0a2c3a29f6b5392e0f8f94ba5669bdc9eb73f6/pkg/controller/nodeipam/ipam/cidr_allocator.go#L41-L46

This PR seems to move the whole nodeipam controller, not only the cloud cidr allocator and I don't think that is correct, since there is a lot of consumers of PodCIDR and the range allocator

/cc @thockin

File: test/integration/ipamperf/results.go [hunk 2]
```diff
@@ -29,8 +29,8 @@ import (
 	"k8s.io/client-go/tools/cache"
 	cloudprovider "k8s.io/cloud-provider"
 	"k8s.io/klog/v2"
-	"k8s.io/kubernetes/pkg/controller/nodeipam/ipam"
```
Comment (aojea) [inline test/integration/ipamperf/results.go:32]: the controller is used out of gce,

**Comment by aojea** [comment]
/sig network
/hold
just for discussion in sig-network too, this is the controller that populates PodCIDR in the node and this was always a hot topic of sig-network

**Comment by aojea** [comment]
should we close?
 there is also a KEP for replacing it

**Pull request closed without merging**